apps/root/src/pages/home/components/portfolio/index.tsx

```diff
@@ -317,44 +316,46 @@ const Portfolio = ({ selectedWalletOption }: PortfolioProps) => {
-
     // Merge multi-chain tokens
-    allTokenKeys.forEach((balanceItemKey) => {
-      const balanceItem = tokenBalances[balanceItemKey];
-      if (!balanceItem) {
-        return;
+    const multiChainTokenBalances = Object.values(balanceTokens).reduce<BalanceItem[]>((acc, balanceToken) => {
+      const equivalentTokenIndex = acc.findIndex((item) =>
+        item.tokens.some((itemToken) => getIsSameOrTokenEquivalent(itemToken.token, balanceToken.token))
+      );
+
+      if (equivalentTokenIndex === -1) {
+        // Unique token
+        acc.push({
+          totalBalanceInUnits: '0',
+          totalBalanceUsd: 0,
+          tokens: [balanceToken],
+          isLoadingPrice: balanceToken.isLoadingPrice,
+          relativeBalance: 0,
+        });
+      } else {
+        // Equivalent token
+        acc[equivalentTokenIndex].tokens.push(balanceToken);
+        // eslint-disable-next-line no-param-reassign
+        acc[equivalentTokenIndex].isLoadingPrice =
+          acc[equivalentTokenIndex].isLoadingPrice || balanceToken.isLoadingPrice;
       }
 
-      const { token: firstAddedToken } = balanceItem.tokens[0];
-
-      firstAddedToken.chainAddresses.forEach(({ address, chainId }) => {
-        const equivalentItemKey = `${chainId}-${address}`;
-        const equivalentItem = tokenBalances[equivalentItemKey];
-        if (equivalentItem && equivalentItemKey !== balanceItemKey) {
-          balanceItem.isLoadingPrice = balanceItem.isLoadingPrice || equivalentItem.isLoadingPrice;
-          const equivalentToken = equivalentItem.tokens[0];
-          balanceItem.tokens.push(equivalentToken);
+      return acc;
+    }, []);
 
-          tokenBalances[balanceItemKey] = balanceItem;
-          delete tokenBalances[`${chainId}-${address}`];
-        }
-      });
-
-      const totalBalanceInUnits = balanceItem.tokens.reduce((acc, { balance, token: equivalentToken }) => {
-        return acc + Number(formatUnits(balance, equivalentToken.decimals));
+    // Calculate totals
+    const multiChainTokenBalancesWithTotal = multiChainTokenBalances.map((balanceItem) => {
+      const totalBalanceInUnits = balanceItem.tokens.reduce((acc, { balance, token }) => {
+        return acc + Number(formatUnits(balance, token.decimals));
       }, 0);
       const totalBalanceUsd = balanceItem.tokens.reduce((acc, { balanceUsd }) => acc + (balanceUsd || 0), 0);
-
       const totalBalanceInUnitsFormatted = parseExponentialNumberToString(totalBalanceInUnits);
 
-      balanceItem.totalBalanceInUnits = totalBalanceInUnitsFormatted;
-      balanceItem.totalBalanceUsd = totalBalanceUsd;
+      return { ...balanceItem, totalBalanceInUnits: totalBalanceInUnitsFormatted, totalBalanceUsd };
     });
 
-    const mappedBalances = map(Object.entries(tokenBalances), ([key, value]) => ({
+    const mappedBalances = map(multiChainTokenBalancesWithTotal, (value, index) => ({
       ...value,
-      key,
+      key: index,
```

Comment:
I don't like this key a lot, can we maybe create it based on the tokens addresses?